### PR TITLE
tags converters: return nil if src is nil

### LINKED ISSUE #2801: Tag converters return map[string]*string instead of nil when the source is nil
/kind bug

**What steps did you take and what happened:**

Tag converters should return `nil` when the source is `nil`. Instead the function returns an empty `map[string]*string` type. You see the following snippets `return tags` where `tags` cannot be `nil`.

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/2349883d1effe6344690459b40e5dc26138010df/azure/converters/tags.go#L24-L33

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/2349883d1effe6344690459b40e5dc26138010df/azure/converters/tags.go#L35-L44

**What did you expect to happen:**

Return `nil` when the source is `nil`.

### PR DESCRIPTION
**Pull request by zioproto**
Fixes #2801 
/kind bug

these converter functions had a bug where the conversion of a nil source was returning an empty `map[string]*string`
this behaviour triggered bugs when doing diff operations on converted objects

This is where this problem was encountered the first time:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2745/files#r1012288112

Release notes:
```release-note
Fixed tags converters functions MapToTags and TagsToMap to return nil on nil input. Added unit tests for tag converters.
```

### CODE REVIEW

File: azure/converters/tags.go [hunk 1]
```diff
@@ -23,6 +23,10 @@ import (
 
 // MapToTags converts a map[string]*string into a infrav1.Tags.
 func MapToTags(src map[string]*string) infrav1.Tags {
+	if src == nil {
```
Comment (CecileRobertMichon) [inline azure/converters/tags.go:26]: can you please add a unit test that covers this scenario?

Comment (zioproto) [reply azure/converters/tags.go:26]: @CecileRobertMichon I now added unit tests for `MapToTags` and `TagsToMap`. Unit tests were missing completely for tag converters if I looked in the correct files, so I created a completely new file. Please review. thanks

**Comment by mboersma** [comment]
When I grep the code for usages of `MapToTags`, I find several places where code seems to expect a non-`nil` return value:

```golang
func (s *Service) isResourceManaged(tags map[string]*string) bool {
	return converters.MapToTags(tags).HasOwned(s.Scope.ClusterName())
}
```

We should add `nil` checks or other appropriate refactoring to account for this change.

**Comment by jackfrancis** [comment]
> When I grep the code for usages of `MapToTags`, I find several places where code seems to expect a non-`nil` return value:
> 
> ```go
> func (s *Service) isResourceManaged(tags map[string]*string) bool {
> 	return converters.MapToTags(tags).HasOwned(s.Scope.ClusterName())
> }
> ```
> 
> We should add `nil` checks or other appropriate refactoring to account for this change.

Would returning a zero value tags object (`infrav1.Tags{}`) be a better choice here instead of returning `nil`?

**Comment by nawazkh** [comment]
> > When I grep the code for usages of `MapToTags`, I find several places where code seems to expect a non-`nil` return value:
> > ```go
> > func (s *Service) isResourceManaged(tags map[string]*string) bool {
> > 	return converters.MapToTags(tags).HasOwned(s.Scope.ClusterName())
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > We should add `nil` checks or other appropriate refactoring to account for this change.
> 
> Would returning a zero value tags object (`infrav1.Tags{}`) be a better choice here instead of returning `nil`?

Returning a zero value tags object (`infrav1.Tags{}`) might lead to throttling again as [being discussed here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2745#discussion_r1012288112)?

**Comment by zioproto** [comment]
I started this PR because the assumption was "it should return nil if the src is nil":
https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2745#discussion_r1017067725

Before doing additional changes I need to understand if this assumption is correct.

@CecileRobertMichon ☝️ 

Thanks :)

**Comment by CecileRobertMichon** [comment]
> We should add nil checks or other appropriate refactoring to account for this change.

+1

Let's audit the code to make sure we are nil checking places where the func is used as needed. 

> Would returning a zero value tags object (infrav1.Tags{}) be a better choice here instead of returning nil?

This is what we're currently doing and that is leading to inconsistencies when the src is nil as discussed [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2745#discussion_r1017067725). Since the function's role is to convert a map variable that can have a nil value, IMO it's a bug to not return nil when the src variable is nil, and the functions in the codebase relying on that faulty behavior should be updated.

File: azure/converters/tags_test.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/onsi/gomega"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+)
+
+func Test_TagsToMap(t *testing.T) {
+	cases := []struct {
+		tags   infrav1.Tags
+		expect map[string]*string
+	}{
+		{
```
Comment (CecileRobertMichon) [inline azure/converters/tags_test.go:32]: let's add a bit more unit test cases to cover other scenarios when the src isn't nil as well

Comment (CecileRobertMichon) [reply azure/converters/tags_test.go:32]: maybe a good test would be to convert the value to map then back to tags and make sure that the value we get is strictly equal to the original value (ie. round trip conversion), since that's the bug we're fixing here.

**Comment by sonasingh46** [comment]
If we are planning to return a nil for nil then we should also keep a track of this method here. 
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/api/v1beta1/tags.go#L39

For example here, 
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/services/privatedns/privatedns.go#L149
We are never passing a nil to the `MapsToTags` method so it is safe to call `HasOwned`

But here,
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/services/groups/groups.go#L133
If a nil tag gets passed to `MapsToTags` then the call to `HasOwned` will panic if  `MapsToTags` returns a nil.

**Comment by sonasingh46** [comment]
> Returning a zero value tags object (infrav1.Tags{}) might lead to throttling
Also @nawazkh can you please help me grok this.

**Comment by nojnhuh** [comment]
> When I grep the code for usages of `MapToTags`, I find several places where code seems to expect a non-`nil` return value:
> 
> ```go
> func (s *Service) isResourceManaged(tags map[string]*string) bool {
> 	return converters.MapToTags(tags).HasOwned(s.Scope.ClusterName())
> }
> ```
> 
> We should add `nil` checks or other appropriate refactoring to account for this change.

FWIW, calling a method on or accessing a nil map won't necessarily panic. e.g. https://go.dev/play/p/-iu_DXwnIrE

So we may not need to change much (if anything). Unit test coverage to verify would be good though.

**Comment by sonasingh46** [comment]
@nojnhuh -- I believe it will panic as the method is value receiver.

**Comment by sonasingh46** [comment]
Ok.. My bad. I don't why I am under impression that when method is value receiver and called over a nil pointer, it panics. ( Could be wrong though )

**Comment by zioproto** [comment]
I pushed the following additional changes:

* nil checking where tag converters are used
* Implemented round trip conversion unit test

I did not squash commits yet to make it easier to audit changes in this PR and eventually rollback individual commits.

Please review and give me feedback when I can squash commits.

Thank you !

File: azure/converters/tags_test.go [hunk 3]
```diff
@@ -75,3 +75,61 @@ func Test_MapToTags(t *testing.T) {
...
+// and make sure that the value we get is strictly equal
+// to the original value (ie. round trip conversion)
+func Test_TagsToMapRoundTrip(t *testing.T) {
+	cases := []struct {
+		tags   infrav1.Tags
+		expect infrav1.Tags
+	}{
+		{
+			tags:   nil,
+			expect: nil,
```
Comment (sonasingh46) [inline azure/converters/tags_test.go:89]: Also worth to add one more case that has empty tags. 
```go
{

tags:   map[string]*string{},
expect: map[string]*string{},

}

```

**Comment by sonasingh46** [comment]
@zioproto -- You can squash the commits. Looks good to me.

**Comment by zioproto** [comment]
@sonasingh46 commits squashed and rebased on current `origin/main`. Please apply the lgtm label again because the automation removed it with the squash

**Review by sonasingh46 — APPROVED** [review]

File: api/v1alpha3/types.go [hunk 4]
```diff
@@ -87,6 +87,9 @@ type VnetSpec struct {
 
 // IsManaged returns true if the vnet is managed.
 func (v *VnetSpec) IsManaged(clusterName string) bool {
+	if v.Tags == nil {
```
Comment (CecileRobertMichon) [inline api/v1alpha3/types.go:90]: I don't think this is needed as @nojnhuh pointed out

` Tags`  is a `map[string]string` , not a pointer to a map, so even if the map is nil it won't panic:

https://go.dev/play/p/-iu_DXwnIrE

File: azure/converters/tags_test.go (hunk 2 above)
Comment (CecileRobertMichon) [reply azure/converters/tags_test.go:32]: thank you for adding those!

File: azure/converters/vm.go [hunk 5]
```diff
@@ -61,7 +61,7 @@ func SDKToVM(v compute.VirtualMachine) *VM {
 		vm.AvailabilityZone = to.StringSlice(v.Zones)[0]
 	}
 
-	if len(v.Tags) > 0 {
+	if v.Tags != nil && len(v.Tags) > 0 {
```
Comment (CecileRobertMichon) [inline azure/converters/vm.go:64]: why do we need this extra check? a nil v.Tags will already not pass the length check https://go.dev/play/p/QoprOQrAlzV

**Comment by CecileRobertMichon** [comment]
> FWIW, calling a method on or accessing a nil map won't necessarily panic. e.g. https://go.dev/play/p/-iu_DXwnIrE

@zioproto, @nojnhuh is correct that calling a method on a nil map doesn't panic so I don't think we need all the extra nil pointer checks. Thanks for adding all the unit tests.

**Comment by zioproto** [comment]
@nojnhuh @CecileRobertMichon I pushed a new commit to address the latest review comments

**Comment by zioproto** [comment]
@sonasingh46 @CecileRobertMichon please take a look again. It should be ready for merge now. Thanks

**Comment by CecileRobertMichon** [comment]
/lgtm
/assign @sonasingh46 @nojnhuh 

@zioproto this is missing a release note

**Review by sonasingh46 — APPROVED** [review]
/lgtm

**Comment by nawazkh** [comment]
Looks good. Thanks!

**Comment by zioproto** [comment]
Release note added, thanks

**Comment by CecileRobertMichon** [comment]
/retest

Error curling Helm 👀
